File: 2024/07/solution_raw.js
const fs = require('fs')

function readInput() {
  const file = fs.readFileSync(process.argv[2]).toString()
  return file.split('\n')
}

function isValid(res, ops) {
  if (Math.floor(res) !== res) {
    return false
  }
  if (ops.length === 1) {
    return res === ops[0]
  }
  return (
    isValid(res - ops[ops.length - 1], ops.slice(0, ops.length - 1)) ||
    isValid(res / ops[ops.length - 1], ops.slice(0, ops.length - 1))
  )
}

function isValid2(res, ops) {
  if (Math.floor(res) !== res) {
    return false
  }
  if (ops.length === 1) {
    return res === ops[0]
  }

  const last = ops[ops.length - 1]
  const resStr = `${res}`

  return (
    isValid2(res - ops[ops.length - 1], ops.slice(0, ops.length - 1)) ||
    isValid2(res / ops[ops.length - 1], ops.slice(0, ops.length - 1)) ||
    (resStr.endsWith(last.toString()) &&
      isValid2(
        +resStr.substring(0, resStr.length - `${last}`.length),
        ops.slice(0, ops.length - 1),
      ))
  )
}

function part1(inp) {
  let res = 0
  for (const line of inp) {
    const [ans, ops] = line.split(': ')

    const opVals = ops.split(' ').map((x) => parseInt(x))
    if (isValid(parseInt(ans), opVals)) {
      res += parseInt(ans)
    }
  }

  return res
}

function part2(inp) {
  let res = 0
  for (const line of inp) {
    const [ans, ops] = line.split(': ')

    const opVals = ops.split(' ').map((x) => parseInt(x))
    if (isValid2(parseInt(ans), opVals)) {
      res += parseInt(ans)
    }
  }

  return res
}

const inp = readInput()
console.log('Part 1:', part1(inp))
console.log('Part 2:', part2(inp))

File: 2025/01/solution_raw.py
import sys

inp = list(l.strip() for l in sys.stdin.readlines())


def part1(inp):
    cur = 50
    ans = 0

    for line in inp:
        d, n = line[0], int(line[1:])
        if d == 'L':
            cur = (cur - n) % 100
        else:
            cur = (cur + n) % 100
        if cur == 0:
            ans += 1
    return ans


def part2(inp):
    cur = 50
    ans = 0

    for line in inp:
        d, n = line[0], int(line[1:])
        val = 1 if d == 'R' else -1
        for i in range(n):
            cur = (cur + val) % 100
            if cur == 0:
                ans += 1

    return ans

print('part1:', part1(inp))
print('part2:', part2(inp))
